File: app/src/main/java/com/cyl/musiclake/ui/music/playlist/PlaylistDetailActivity.java
package com.cyl.musiclake.ui.music.playlist;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.InputType;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ImageView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.cyl.musiclake.R;
import com.cyl.musiclake.RxBus;
import com.cyl.musiclake.base.BaseActivity;
import com.cyl.musiclake.common.Constants;
import com.cyl.musiclake.common.Extras;
import com.cyl.musiclake.data.PlayHistoryLoader;
import com.cyl.musiclake.data.db.Music;
import com.cyl.musiclake.data.db.Playlist;
import com.cyl.musiclake.db.Album;
import com.cyl.musiclake.db.Artist;
import com.cyl.musiclake.event.PlaylistEvent;
import com.cyl.musiclake.player.PlayManager;
import com.cyl.musiclake.ui.OnlinePlaylistUtils;
import com.cyl.musiclake.ui.UIUtilsKt;
import com.cyl.musiclake.ui.music.dialog.PopupDialogFragment;
import com.cyl.musiclake.ui.music.local.adapter.SongAdapter;
import com.cyl.musiclake.ui.zone.EditActivity;
import com.cyl.musiclake.utils.CoverLoader;
import com.cyl.musiclake.utils.LogUtil;
import com.cyl.musiclake.view.ItemDecoration;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.OnClick;

/**
 * 作者：yonglong on 2016/8/15 19:54
 * 邮箱：643872807@qq.com
 * 版本：2.5
 */
public class PlaylistDetailActivity extends BaseActivity<PlaylistDetailPresenter> implements PlaylistDetailContract.View {

    @BindView(R.id.recyclerView)
    RecyclerView mRecyclerView;
    @BindView(R.id.toolbar)
    Toolbar mToolbar;
    @BindView(R.id.album_art)
    ImageView album_art;
    @BindView(R.id.fab)
    FloatingActionButton mFab;

    @OnClick(R.id.fab)
    void onPlayAll() {
        PlayManager.play(0, musicList, mPlaylist.getPid());
    }

    private SongAdapter mAdapter;
    private List<Music> musicList = new ArrayList<>();
    private Playlist mPlaylist;
    private Artist mArtist;
    private Album mAlbum;
    private String title;
    private String coverUrl;

    public static void newInstance(Context context, Playlist playlist) {
        Intent intent = new Intent(context, PlaylistDetailActivity.class);
        intent.putExtra(Extras.PLAYLIST, playlist);
        context.startActivity(intent);
    }

    public static void newInstance(Context context, Artist artist) {
        Intent intent = new Intent(context, PlaylistDetailActivity.class);
        intent.putExtra(Extras.ARTIST, artist);
        context.startActivity(intent);
    }

    public static void newInstance(Context context, Album album) {
        Intent intent = new Intent(context, PlaylistDetailActivity.class);
        intent.putExtra(Extras.ALBUM, album);
        context.startActivity(intent);
    }


    @Override
    protected int getLayoutResID() {
        return R.layout.frag_playlist_detail;
    }


    @Override
    protected void initData() {
        showLoading();
        if (mPlaylist != null && mPresenter != null)
            mPresenter.loadPlaylistSongs(mPlaylist);
        if (mArtist != null && mPresenter != null)
            mPresenter.loadArtistSongs(mArtist);
        if (mAlbum != null && mPresenter != null)
            mPresenter.loadAlbumSongs(mAlbum);
    }

    @Override
    protected String setToolbarTitle() {
        mPlaylist = (Playlist) getIntent().getParcelableExtra(Extras.PLAYLIST);
        mArtist = (Artist) getIntent().getSerializableExtra(Extras.ARTIST);
        mAlbum = (Album) getIntent().getSerializableExtra(Extras.ALBUM);

        if (mPlaylist != null) title = mPlaylist.getName();
        if (mArtist != null) title = mArtist.getName();
        if (mAlbum != null) title = mAlbum.getName();
        return title;
    }

    @Override
    protected void initView() {
        mAdapter = new SongAdapter(musicList);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mRecyclerView.addItemDecoration(new ItemDecoration(this, ItemDecoration.VERTICAL_LIST));
        mRecyclerView.setAdapter(mAdapter);
        mAdapter.bindToRecyclerView(mRecyclerView);
    }

    @Override
    protected void initInjector() {
        mActivityComponent.inject(this);
    }

    @Override
    protected void listener() {
        mAdapter.setOnItemClickListener((adapter, view, position) -> {
            if (view.getId() != R.id.iv_more) {
                if (mPlaylist != null) {
                    PlayManager.play(position, musicList, mPlaylist.getPid());
                } else if (mArtist != null) {
                    PlayManager.play(position, musicList, String.valueOf(mArtist.getId()));
                } else if (mAlbum != null) {
                    PlayManager.play(position, musicList, String.valueOf(mAlbum.getId()));
                }
                mAdapter.notifyDataSetChanged();
            }
        });
        mAdapter.setOnItemChildClickListener((adapter, view, position) -> {
            Music music = musicList.get(position);
            PopupDialogFragment.Companion.newInstance(music)
                    .show(this);
        });
    }

    @SuppressLint("ResourceAsColor")
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.action_delete_playlist:
                LogUtil.e("action_delete_playlist");
                UIUtilsKt.deletePlaylist(PlaylistDetailActivity.this, mPlaylist, isHistory -> {
                    if (isHistory) {
                        musicList.clear();
                        PlayHistoryLoader.INSTANCE.clearPlayHistory();
                        mAdapter.notifyDataSetChanged();
                        showEmptyState();
                        RxBus.getInstance().post(new PlaylistEvent(Constants.PLAYLIST_HISTORY_ID));
                    } else if (mPresenter != null) {
                        OnlinePlaylistUtils.INSTANCE.deletePlaylist(mPlaylist, result -> {
                            onBackPress();
                            return null;
                        });
                    }
                    return null;
                }, () -> null);
                break;
            case R.id.action_rename_playlist:
                LogUtil.e("action_rename_playlist");
                new MaterialDialog.Builder(this)
                        .title("重命名歌单")
                        .positiveText("确定")
                        .negativeText("取消")
                        .inputRangeRes(2, 10, R.color.red)
                        .inputType(InputType.TYPE_CLASS_TEXT)
                        .input("输入歌单名", mPlaylist.getName(), false, (dialog, input) -> LogUtil.e("=====", input.toString()))
                        .onPositive((dialog, which) -> {
                            String title = dialog.getInputEditText().getText().toString();
                            mPresenter.renamePlaylist(mPlaylist, title);
                        })
                        .positiveText("确定")
                        .negativeText("取消")
                        .show();
                break;
            case R.id.action_share:
                Intent intent3 = new Intent(this, EditActivity.class);
                StringBuilder content = new StringBuilder();
                if (musicList.size() > 0) {
                    content = new StringBuilder("分享歌单\n");
                }
                for (int i = 0; i < musicList.size(); i++) {
                    content.append(musicList.get(i).getTitle()).append("---").append(musicList.get(i).getArtist());
                    content.append("\n");
                }
                intent3.putExtra("content", content.toString());
                startActivity(intent3);
                break;
        }
        return super.

                onOptionsItemSelected(item);

    }

    private void onBackPress() {
        this.onBackPressed();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_playlist_detail, menu);
        if (mPlaylist == null) {
            menu.removeItem(R.id.action_rename_playlist);
            menu.removeItem(R.id.action_delete_playlist);
        } else if (mPlaylist.getPid() != null && mPlaylist.getPid().equals(Constants.PLAYLIST_HISTORY_ID)) {
            menu.removeItem(R.id.action_rename_playlist);
        } else if (mPlaylist.getPid() != null && mPlaylist.getPid().equals(Constants.PLAYLIST_LOVE_ID)) {
            menu.removeItem(R.id.action_rename_playlist);
            menu.removeItem(R.id.action_delete_playlist);
        } else if (mPlaylist.getType() == 2) {
            menu.removeItem(R.id.action_rename_playlist);
            menu.removeItem(R.id.action_delete_playlist);
        }
        return super.onCreateOptionsMenu(menu);
    }


    @Override
    public void showPlaylistSongs(List<Music> songList) {
        hideLoading();
        musicList.addAll(songList);
        mAdapter.setNewData(musicList);
        if (mPlaylist != null && mPlaylist.getCoverUrl() != null) {
            coverUrl = mPlaylist.getCoverUrl();
        } else if (mArtist != null && mArtist.getPicUrl() != null) {
            coverUrl = mArtist.getPicUrl();
        } else if (musicList.size() >= 1) {
            coverUrl = musicList.get(0).getCoverUri();
        }
        CoverLoader.loadImageView(getContext(), coverUrl, album_art);
        if (musicList.size() == 0) {
            showEmptyState();
        }
    }

    @Override
    public void changePlayStatus(Boolean isPlaying) {

    }

    @Override
    public void removeMusic(int position) {
        musicList.remove(position);
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public void success(int type) {
        onBackPress();
    }
}

File: app/src/main/java/com/cyl/musiclake/ui/music/playlist/PlaylistDetailActivity.kt
package com.cyl.musiclake.ui.music.playlist

import android.support.v7.widget.LinearLayoutManager
import android.text.InputType
import android.view.Menu
import android.view.MenuItem
import com.afollestad.materialdialogs.MaterialDialog
import com.cyl.musiclake.R
import com.cyl.musiclake.base.BaseActivity
import com.cyl.musiclake.bean.Album
import com.cyl.musiclake.bean.Artist
import com.cyl.musiclake.bean.Music
import com.cyl.musiclake.bean.Playlist
import com.cyl.musiclake.common.Constants
import com.cyl.musiclake.common.Extras
import com.cyl.musiclake.common.NavigationHelper
import com.cyl.musiclake.data.PlayHistoryLoader
import com.cyl.musiclake.event.PlaylistEvent
import com.cyl.musiclake.player.PlayManager
import com.cyl.musiclake.ui.OnlinePlaylistUtils
import com.cyl.musiclake.ui.deletePlaylist
import com.cyl.musiclake.ui.music.dialog.BottomDialogFragment
import com.cyl.musiclake.ui.music.edit.EditSongListActivity
import com.cyl.musiclake.ui.music.local.adapter.SongAdapter
import com.cyl.musiclake.utils.CoverLoader
import com.cyl.musiclake.utils.LogUtil
import com.cyl.musiclake.view.ItemDecoration
import kotlinx.android.synthetic.main.frag_playlist_detail.*
import kotlinx.android.synthetic.main.fragment_recyclerview_notoolbar.*
import org.greenrobot.eventbus.EventBus
import org.jetbrains.anko.startActivity

/**
 * 作者：yonglong on 2016/8/15 19:54
 * 邮箱：643872807@qq.com
 * 版本：2.5
 */
class PlaylistDetailActivity : BaseActivity<PlaylistDetailPresenter>(), PlaylistDetailContract.View, BottomDialogFragment.RemoveMusicListener {
    private var mAdapter: SongAdapter? = null
    private val musicList = mutableListOf<Music>()
    private var mPlaylist: Playlist? = null
    private var mArtist: Artist? = null
    private var pid: String? = null
    private var mAlbum: Album? = null
    private var title: String? = null
    private var coverUrl: String? = null

    private var bottomDialogFragment: BottomDialogFragment? = null

    override fun getLayoutResID(): Int {
        return R.layout.frag_playlist_detail
    }

    override fun initData() {
        showLoading()
        mPlaylist?.let {
            if (it.musicList.size > 0) {
                showPlaylistSongs(it.musicList)
            } else {
                mPresenter?.loadPlaylistSongs(it)
            }
        }
        mArtist?.let {
            mPresenter?.loadArtistSongs(it)
        }
        mAlbum?.let {
            mPresenter?.loadAlbumSongs(it)
        }
    }

    override fun setToolbarTitle(): String? {
        mPlaylist = intent.getParcelableExtra(Extras.PLAYLIST)
        mArtist = intent.getParcelableExtra(Extras.ARTIST)
        mAlbum = intent.getSerializableExtra(Extras.ALBUM) as Album?
        mPlaylist?.let {
            title = it.name
            pid = it.pid
            coverUrl = it.coverUrl
            CoverLoader.loadImageView(context, coverUrl, album_art)
        }
        mArtist?.let {
            title = it.name
            pid = it.artistId.toString()
            coverUrl = it.picUrl
            CoverLoader.loadImageView(context, coverUrl, album_art)
        }
        mAlbum?.let {
            title = it.name
            pid = it.albumId.toString()
            coverUrl = it.cover
            CoverLoader.loadImageView(context, coverUrl, album_art)
        }
        return title
    }

    override fun initView() {
        mAdapter = SongAdapter(musicList)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.addItemDecoration(ItemDecoration(this, ItemDecoration.VERTICAL_LIST))
        recyclerView.adapter = mAdapter
        mAdapter?.bindToRecyclerView(recyclerView)
        fab.setOnClickListener { PlayManager.play(0, musicList, pid) }
    }

    override fun initInjector() {
        mActivityComponent.inject(this)
    }

    override fun listener() {
        mAdapter?.setOnItemClickListener { _, view, position ->
            if (view.id != R.id.iv_more) {
                when {
                    mPlaylist != null -> PlayManager.play(position, musicList, mPlaylist?.pid)
                    mArtist != null -> PlayManager.play(position, musicList, mArtist?.artistId.toString())
                    mAlbum != null -> PlayManager.play(position, musicList, mAlbum?.albumId.toString())
                }
                mAdapter?.notifyDataSetChanged()
                NavigationHelper.navigateToPlaying(this, view.findViewById(R.id.iv_cover))
            }
        }
        mAdapter?.setOnItemChildClickListener { _, _, position ->
            val music = musicList[position]
            if (mPlaylist != null && mPlaylist?.type == Playlist.PT_MY) {
                bottomDialogFragment = BottomDialogFragment.newInstance(music, Constants.OP_PLAYLIST)
                bottomDialogFragment?.removeMusicListener = this
                bottomDialogFragment?.position = position
                bottomDialogFragment?.show(this)
            } else {
                bottomDialogFragment = BottomDialogFragment.newInstance(music, Constants.OP_ONLINE)
                bottomDialogFragment?.position = position
                bottomDialogFragment?.show(this)
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        when (id) {
            R.id.action_delete_playlist -> {
                LogUtil.e("action_delete_playlist")
                mPlaylist?.let {
                    deletePlaylist(it, success = { isHistory ->
                        if (isHistory) {
                            musicList.clear()
                            PlayHistoryLoader.clearPlayHistory()
                            mAdapter?.notifyDataSetChanged()
                            showEmptyState()
                            EventBus.getDefault().post(PlaylistEvent(Constants.PLAYLIST_HISTORY_ID))
                        } else if (mPresenter != null) {
                            OnlinePlaylistUtils.deletePlaylist(it) { _ ->
                                onBackPress()
                            }
                        }
                    })
                }
            }
            R.id.action_rename_playlist -> {
                LogUtil.e("action_rename_playlist")
                mPlaylist?.let {
                    MaterialDialog.Builder(this)
                            .title(R.string.playlist_rename)
                            .positiveText(R.string.sure)
                            .negativeText(R.string.cancel)
                            .inputRangeRes(2, 10, R.color.red)
                            .inputType(InputType.TYPE_CLASS_TEXT)
                            .input("输入歌单名", it.name, false) { dialog, input -> LogUtil.e("=====", input.toString()) }
                            .onPositive { dialog, _ ->
                                val title = dialog.inputEditText?.text.toString()
                                if (title != mPlaylist?.name) {
                                    mPresenter?.renamePlaylist(it, title)
                                }
                            }
                            .positiveText(R.string.sure)
                            .negativeText(R.string.cancel)
                            .show()
                }
            }
            R.id.action_batch -> {
                startActivity<EditSongListActivity>(Extras.SONG_LIST to musicList)
            }
        }
        return super.onOptionsItemSelected(item)

    }

    private fun onBackPress() {
        this.onBackPressed()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_playlist_detail, menu)
        if (mPlaylist == null) {
            menu.removeItem(R.id.action_rename_playlist)
            menu.removeItem(R.id.action_delete_playlist)
        }
        mPlaylist?.let {
            when (it.pid) {
                Constants.PLAYLIST_HISTORY_ID -> {
                    menu.removeItem(R.id.action_rename_playlist)
                }
                Constants.PLAYLIST_LOVE_ID -> {
                    menu.removeItem(R.id.action_rename_playlist)
                    menu.removeItem(R.id.action_delete_playlist)
                }
            }
            if (it.type == Playlist.PT_BAIDU) {
                //百度电台
                menu.removeItem(R.id.action_rename_playlist)
                menu.removeItem(R.id.action_delete_playlist)
            }
        }

        return super.onCreateOptionsMenu(menu)
    }


    override fun showPlaylistSongs(songList: MutableList<Music>?) {
        hideLoading()
        songList?.let {
            musicList.addAll(songList)
        }
        mAdapter?.setNewData(musicList)
        if (coverUrl == null) {
            mPlaylist?.coverUrl?.let {
                coverUrl = it
            }
            mArtist?.picUrl?.let {
                coverUrl = it
            }
            if (musicList.size > 0 && coverUrl == null) {
                coverUrl = musicList[0].coverUri
            }
            CoverLoader.loadImageView(context, coverUrl, album_art)
        }
        if (musicList.size == 0) {
            showEmptyState()
        }
    }

    override fun removeMusic(position: Int) {
        musicList.removeAt(position)
        mAdapter?.notifyDataSetChanged()
        if (musicList.size == 0) {
            showEmptyState()
        }
    }

    override fun success(type: Int) {
        onBackPress()
    }

    /**
     * 移除歌手歌曲
     */
    override fun remove(position: Int, music: Music?) {
        OnlinePlaylistUtils.disCollectMusic(pid, music) {
            removeMusic(position)
        }
    }

    /**
     * 点击重试按钮响应事件
     */
    override fun retryLoading() {
        super.retryLoading()
        showLoading()
        mPlaylist?.let {
            if (it.musicList.size > 0) {
                showPlaylistSongs(it.musicList)
            } else {
                mPresenter?.loadPlaylistSongs(it)
            }

        }
        mArtist?.let {
            mPresenter?.loadArtistSongs(it)
        }
        mAlbum?.let {
            mPresenter?.loadAlbumSongs(it)
        }
    }
}
